File: backend/src/test/java/ca/metricalsky/winston/entity/fetch/FetchOperationEntityTest.java
package ca.metricalsky.winston.entity.fetch;

import ca.metricalsky.winston.entity.fetch.FetchOperationEntity.Status;
import ca.metricalsky.winston.entity.fetch.FetchOperationEntity.Type;
import ca.metricalsky.winston.test.TestUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;

import static org.assertj.core.api.Assertions.assertThat;


@DataJpaTest
class FetchOperationEntityTest {

    @Autowired
    private TestEntityManager entityManager;

    @Test
    void persistsWithOnlyRequiredFields() {
        var fetchRequestEntity = persistFetchRequest();
        var fetchOperationEntity = FetchOperationEntity.builder()
                .fetchRequestId(fetchRequestEntity.getId())
                .operationType(Type.CHANNELS)
                .objectId(TestUtils.randomId())
                .build();

        var persistedEntity = entityManager.persistFlushFind(fetchOperationEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("fetchRequestId", fetchOperationEntity.getFetchRequestId())
                .hasFieldOrPropertyWithValue("operationType", fetchOperationEntity.getOperationType())
                .hasFieldOrPropertyWithValue("objectId", fetchOperationEntity.getObjectId())
                .hasFieldOrPropertyWithValue("status", Status.READY);
        assertThat(persistedEntity.getId())
                .isNotNull();
        assertThat(persistedEntity.getCreatedAt())
                .isNotNull();
        assertThat(persistedEntity.getLastUpdatedAt())
                .isNotNull();
    }

    @Test
    void persistsWithAllOptionalFields() {
        var fetchRequestEntity = persistFetchRequest();
        var fetchOperationEntity = FetchOperationEntity.builder()
                .fetchRequestId(fetchRequestEntity.getId())
                .operationType(Type.CHANNELS)
                .objectId(TestUtils.randomId())
                .mode(TestUtils.randomString())
                .publishedAfter(OffsetDateTime.now())
                .publishedBefore(OffsetDateTime.now())
                .status(Status.FAILED)
                .error(TestUtils.randomString())
                .build();

        var persistedEntity = entityManager.persistFlushFind(fetchOperationEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("fetchRequestId", fetchOperationEntity.getFetchRequestId())
                .hasFieldOrPropertyWithValue("operationType", fetchOperationEntity.getOperationType())
                .hasFieldOrPropertyWithValue("objectId", fetchOperationEntity.getObjectId())
                .hasFieldOrPropertyWithValue("mode", fetchOperationEntity.getMode())
                .hasFieldOrPropertyWithValue("publishedAfter",
                        fetchOperationEntity.getPublishedAfter().withOffsetSameInstant(ZoneOffset.UTC))
                .hasFieldOrPropertyWithValue("publishedBefore",
                        fetchOperationEntity.getPublishedBefore().withOffsetSameInstant(ZoneOffset.UTC))
                .hasFieldOrPropertyWithValue("status", fetchOperationEntity.getStatus())
                .hasFieldOrPropertyWithValue("error", fetchOperationEntity.getError())
                .hasNoNullFieldsOrProperties();
    }

    private FetchRequestEntity persistFetchRequest() {
        var fetchRequestEntity = FetchRequestEntity.builder()
                .build();
        return entityManager.persist(fetchRequestEntity);
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/entity/fetch/FetchOperationEntityTest.kt
package ca.metricalsky.winston.database.entity.fetch

import ca.metricalsky.winston.database.entity.fetch.FetchOperationEntity.Status.READY
import ca.metricalsky.winston.database.test.annotation.DatabaseTest
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager

@DatabaseTest
class FetchOperationEntityTest(
    entityManager: TestEntityManager,
): WordSpec({

    val faker = DatabaseFaker();

    "FetchOperationEntity" should {

        "persist with only required fields" {
            val fetchOperation = faker.fetchOperation().minimalEntity()

            entityManager.persistFlushFind(fetchOperation) shouldNotBeNull {
                id shouldNotBe null
                operationType shouldBe fetchOperation.operationType
                objectId shouldBe fetchOperation.objectId
                status shouldBe READY
                createdAt shouldNotBe null
                lastUpdatedAt shouldNotBe null
            }
        }

        "persist with all optional fields" {
            val fetchRequest = entityManager.persist(faker.fetchRequest().minimalEntity())
            val fetchOperation = faker.fetchOperation().completeEntity(fetchRequest)

            entityManager.persistFlushFind(fetchOperation) shouldNotBeNull {
                id shouldNotBe null
                fetchRequestId shouldBe fetchRequest.id
                operationType shouldBe fetchOperation.operationType
                objectId shouldBe fetchOperation.objectId
                mode shouldBe fetchOperation.mode
                publishedAfter shouldBe fetchOperation.publishedAfter
                publishedBefore shouldBe fetchOperation.publishedBefore
                status shouldBe fetchOperation.status
                createdAt shouldNotBe null
                lastUpdatedAt shouldNotBe null
            }
        }
    }
})
